rectools/models/sasrec.py

```diff
@@ -1069,41 +1017,199 @@ def _calc_softmax_loss(cls, logits: torch.Tensor, y: torch.Tensor, w: torch.Tens
-        return loss
+        """
+        Forward pass. Returns calculated loss.
+
+        Parameters
+        ----------
+        batch: tp.Dict[str, torch.Tensor]
+            Full batch from train_dataloader. Different aspects (like x or y) can be accessed with corresponding keys.
+        item_embs: torch.Tensor
+            Full catalog item embeddings from torch_model.forward method.
+        session_embs: torch.Tensor
+            Session embeddings from torch_model.forward method.
+        """
+        logits = self._get_logits(batch, item_embs, session_embs)
+
+        # Transform logits before calculating BCE loss
+        n_negatives = batch["negatives"].shape[2]
+        logits = self._get_reduced_overconfidence_logits(logits, self.n_actual_items, n_negatives, self.gbce_t)
+
+        return self._calc_loss(batch, logits)
+
+    def _get_reduced_overconfidence_logits(
+        self, logits: torch.Tensor, n_actual_items: int, n_negatives: int, gbce_t: float
+    ) -> torch.Tensor:
+        """Transform positive items logits according to https://arxiv.org/pdf/2308.07192.pdf."""
+        alpha = n_negatives / (n_actual_items - 1)  # negatives sampling rate
+        beta = alpha * (gbce_t * (1 - 1 / alpha) + 1 / alpha)
+
+        pos_logits = logits[:, :, 0:1].to(torch.float64)
+        neg_logits = logits[:, :, 1:].to(torch.float64)
+
+        epsilon = 1e-10
+        pos_probs = torch.clamp(torch.sigmoid(pos_logits), epsilon, 1 - epsilon)
+        pos_probs_adjusted = torch.clamp(pos_probs.pow(-beta), 1 + epsilon, torch.finfo(torch.float64).max)
+        pos_probs_adjusted = torch.clamp(
+            torch.div(1, (pos_probs_adjusted - 1)), epsilon, torch.finfo(torch.float64).max
+        )
+        pos_logits_transformed = torch.log(pos_probs_adjusted)
+        logits = torch.cat([pos_logits_transformed, neg_logits], dim=-1)
+        return logits
+
+
+# ####  --------------  Lightning Module  --------------  #### #
+
+
+class SessionEncoderLightningModuleBase(LightningModule):
+    """
+    Base class for lightning module. To change train procedure inherit
+    from this class and pass your custom LightningModule to transformer-based recommender model during its
+    initialization.
+
+    Parameters
+    ----------
+    torch_model: TransformerBasedSessionEncoder
+        Torch for encoding sessions using transformer architecture.
+    loss: Union[Literal["softmax", "BCE", "gBCE"], SessionEncoderHeadBase]
+        Loss function specification or custom session encoder head implementation.
+    lr: float
+        Learning rate.
+    gbce_t: float
+        Calibration parameter for gBCE loss.
+    n_actual_items: int
+        Number of actual items in catalog. Needed for calculating negatives sampling rate for gBCE loss.
+    """
+
+    def __init__(
+        self,
+        torch_model: TransformerBasedSessionEncoder,
+        loss: tp.Union[LossDescription, SessionEncoderHeadBase],
+        lr: float,
+        gbce_t: float,
```

Comment:
Should be optional, right? Since it's used only for one specific loss

Reply:
It' more convenient to have actual default params for such things. It this case we have out-of-the-box decisions for all loss variants. n_negatives is also needed for 2 losses out of 3

Reply:
But here are no default values...

Also n_negatives I see used only for 1 loss

Reply:
We have default value  for bot `gbce_t` and `n_negatives` in model initialisation. So we always assume it's a float and an int.

n_negatives is used for BCE and gBCE, not used for softmax